kolibri/plugins/coach/assets/src/state/actions/exam.js

```diff
@@ -363,10 +390,72 @@ function fetchContent(store, topicId) {
               return subtopic;
             });
 
+            resolve({ topic, subtopics, exercises });
+          },
+          error => reject(error)
+        );
+      },
+      error => reject(error)
+    );
+  });
+}
+
+function goToTopic(store, topicId) {
+  return new Promise((resolve, reject) => {
+    fetchTopic(store, topicId).then(
+      content => {
+        store.dispatch('SET_TOPIC', content.topic);
+        store.dispatch('SET_SUBTOPICS', content.subtopics);
+        store.dispatch('SET_EXERCISES', content.exercises);
+        resolve();
+      },
+      error => reject(error)
+    );
+  });
+}
+
+// TODO: Optimize
+function goToTopLevel(store) {
+  return new Promise((resolve, reject) => {
+    const channelPromise = ChannelResource.getCollection({ available: true }).fetch();
+
+    ConditionalPromise.all([channelPromise]).only(
+      CoreActions.samePageCheckGenerator(store),
+      ([channelsCollection]) => {
+        const fetchTopicPromises = channelsCollection.map(channel =>
+          fetchTopic(store, channel.root)
+        );
+        ConditionalPromise.all(fetchTopicPromises).only(
+          CoreActions.samePageCheckGenerator(store),
+          channelsContent => {
+            const subtopics = channelsContent.map(channel => {
+              const subtopic = channel.topic;
+              let allExercisesWithinSubtopic = [];
+              channel.subtopics.forEach(subtopic => {
+                allExercisesWithinSubtopic = concat(
+                  allExercisesWithinSubtopic,
+                  subtopic.allExercisesWithinTopic
+                );
+              });
```

Comment:
rather than using `let` and `forEach`, consider using `reduce`.

not tested, but maybe something like:

```javascript
const allExercisesWithinSubtopic  = channel.subtopics.reduce(
  (accumulator, subtopic) => accumulator.concat(subtopic.allExercisesWithinTopic),
  []
);
```